### PR TITLE
feat: update adds improved jsdoc based message

### PR DESCRIPTION
**Pull request by DarkPurple141**
Applies a jsdoc format instead to better explain usage.

### CODE REVIEW

File: src/tiny-invariant.ts [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
-// Throw an error if the condition fails
-// Strip out error messages for production
-// > Not providing an inline default argument for message as the result is smaller
+/**
+ * Throw an error if the condition fails. Strip out error messages for production.
+ *
+ * @param condition A boolean condition - if falsey will thrown an error.
+ * @param message The message provided to accompany the invariant. No inline default argument for message as the result is smaller.
+ *
+ * @example
```
Comment (alexreardon) [inline src/tiny-invariant.ts:10]: I like the `@example`

File: src/tiny-invariant.ts [hunk 2]
```diff
@@ -1,9 +1,20 @@
 const isProduction: boolean = process.env.NODE_ENV === 'production';
 const prefix: string = 'Invariant failed';
 
-// Throw an error if the condition fails
-// Strip out error messages for production
-// > Not providing an inline default argument for message as the result is smaller
+/**
+ * Throw an error if the condition fails. Strip out error messages for production.
+ *
+ * @param condition A boolean condition - if falsey will thrown an error.
```
Comment (alexreardon) [inline src/tiny-invariant.ts:7]: Do we have to provide definitions for the `@param`s? It feels like doubling up 🤔

Comment (alexreardon) [reply src/tiny-invariant.ts:7]: I wonder if this is enough as a `jsdoc`:

```ts
/** Throw an `Error` if the `condition` is falsey */
```

Comment (alexreardon) [reply src/tiny-invariant.ts:7]: What do you think @DarkPurple141 ?

Comment (alexreardon) [inline src/tiny-invariant.ts:7]: Suggestion:

> @param condition - if falsey will `throw` an `Error`.

Comment (DarkPurple141) [reply src/tiny-invariant.ts:7]: Don't need them necessarily, really just there for completeness.

Comment (DarkPurple141) [reply src/tiny-invariant.ts:7]: Could be better yeah

File: src/tiny-invariant.ts (hunk 1 above)
Comment (DarkPurple141) [reply src/tiny-invariant.ts:13]: haha yeah. Playful but I think it helps to make it really clear how to use the fn :)

File: src/tiny-invariant.ts [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+ * ```tsx
+ * import invariant from 'tiny-invariant';
+ *
+ * // throws when 1 no longer equals 1
+ * invariant(1 === 1, 'Maths is broken');
+ * ```
+ */
 export default function invariant(
   condition: any,
-  // Can provide a string, or a function that returns a string for cases where
```
Comment (DarkPurple141) [inline src/tiny-invariant.ts:9]: This is now covered directly in the jsdoc

File: src/tiny-invariant.ts [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
-// > Not providing an inline default argument for message as the result is smaller
+/**
+ * __invariant__
+ * 
+ * The `invariant` function takes a value, and if the value is falsy then the `invariant` function will throw. If the value is truthy, then the function will not throw.
+ *
+ * @param {boolean} condition A boolean condition - if falsey will thrown an error.
+ * @param {(string|() => string)} message The message provided to accompany the invariant. Can provide a string, or a function that returns a string for cases where the message takes a fair amount of effort to compute.
+ *  
+ * @returns {asserts condition is true}
```
Comment (alexreardon) [inline src/tiny-invariant.ts:12]: asserts condition is truthy?

File: src/tiny-invariant.ts [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
 
-// Throw an error if the condition fails
-// Strip out error messages for production
-// > Not providing an inline default argument for message as the result is smaller
+/**
+ * __invariant__
+ * 
+ * The `invariant` function takes a value, and if the value is falsy then the `invariant` function will throw. If the value is truthy, then the function will not throw.
+ *
+ * @param {boolean} condition A boolean condition - if falsey will thrown an error.
```
Comment (alexreardon) [inline src/tiny-invariant.ts:9]: the `param` can be any value

File: src/tiny-invariant.ts (hunk 4 above)
Comment (DarkPurple141) [reply src/tiny-invariant.ts:12]: `{ }` syntax is a hint for the TS server - so this is really
`asserts condition (variable) is true` <-- which is the ts expression

Can certainly add the explanation as an annotation tho next to it.

File: src/tiny-invariant.ts (hunk 5 above)
Comment (DarkPurple141) [reply src/tiny-invariant.ts:9]: yeah this needs to be corrected - will address

**Comment by alexreardon** [comment]
Thanks for your effort on this @DarkPurple141. Rather than re-implementing the types in jsdoc, for now I've decided to keep the types in `TypeScript`, and add jsdoc information to help with clarity for consumers

**Comment by alexreardon** [comment]
#177

**Comment by DarkPurple141** [comment]
Hi Alex thanks - what you've merged is a improvement so I appreciate that it'll make it easier to parse in an IDE!

I'm a bit disappointed that we couldn't merge this PR - I spent a decent amount of time nursing it. The types aren't redeclared, typescript annotations still win as per the attached screen grabs. This did however provide type hints where a typescript server is not present which is not uncommon. Maybe this context was lost in the thread of this PR?

I feel like what's been merged is completely fine but borrows heavily from the context of this contribution.

**Comment by alexreardon** [comment]
Hey Alex, I am sorry I disappointed you - that was not my intention at all. I really appreciate your contribution.

When I accept a contribution, I want to completely understand every line as at the end of the day - I'll be on the hook for it. I haven't looked into jsdoc type assertions, and I would want to check that what was in this PR behaved correctly in non-TS environments before merging. I didn't really have the energy to do that right now. I thought I would cut down the scope of this change (for now) to just be about making editor experiences better. Rather than asking you to modify this PR, I thought it would be easier for everybody for me to write and ship a small jsdoc comment based on the language in the existing documentation.

**Pull request closed without merging**